design_drafts/ros2_control_documentation.rst

```diff
@@ -0,0 +1,273 @@
+
+Differences to ros_control (ROS1)
+=================================
+
+Hardware Structures - classes
+-----------------------------
+
+The ros_control uses the ``RobotHW`` class as a rigid structure to handle any hardware.
+This makes it impossible to extend the existing robot with additional hardware, like sensors, actuators, and tools, without coding.
+
+The ros2_control defines three different types of hardware ``Actuator``, ``Sensor`` and ``System``.
+Using a combination (composition) of those basic components, any physical robotic cell (robot and its surrounding) can be described.
+This also means that multi-robot, robot-sensor, robot-gripper combinations are supported out of the box.
+Section `Hardware Components <#hardware-components>`__ describes this in detail.
+
+Hardware Interfaces
+-------------------
+
+The ros_control allows only three types of interfaces (joints), i.e., ``position``, ``velocity``, and ``effort``. The ``RobotHW`` class makes it very hard to use any other data to control the robot.
+
+The ros2_control does not mandate a fixed set of interface types, but they are defined as strings in `hardware's description <#hardware-description-in-urdf>`__.
+To ensure compatibility of standard controllers, standard interfaces are defined as constants in `hardware_interface package <https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/include/hardware_interface/types/hardware_interface_type_values.hpp>`__.
+
+Controller's Access to Hardware
+-------------------------------
+
+In ros_control, the controllers had direct access to the ``RobotHW`` class requesting access to its interfaces (joints).
+The hardware itself then took care of registered interfaces and resource conflicts.
+
+In ros2_control, ``ResourceManager`` takes care of the state of available interfaces in the framework and enables controllers to access the hardware.
+Also, the controllers do not have direct access to hardware anymore, but they register their interfaces to the `ControllerManager`.
+
+Migration Guide to ros2_control
+===============================
+
+RobotHardware to Components
+---------------------------
+#. Forget your implementation or the ``RobotHW`` interface. This is not used anymore. (Do not delete it, you can still extract some code.)
+#. Decide which component type is suitable for your case. Maybe it makes sense to separate ``RobotHW`` into multiple components.
+#. Implement `ActuatorInterface`_, `SensorInterface`_ or `SystemInterface`_ classes as follows:
+   
+   #. In the constructor, initialize all variables needed for communication with your hardware or define the default one.
+   #. In the configure function, read all the parameters your hardware needs from the parsed URDF snippet (i.e., from the `HardwareInfo`_ structure). Here you can cross-check if all joints and interfaces in URDF have allowed values or a combination of values.
```

Comment:
these components can do their own parsing and are not limited to HardwareInfo

Reply:
Is it so? I am not sure what you mean. We provide all the data in the `HardwareInfo` structure. so you are actually not parsing, but iterating through the parsed structure.

Reply:
You are right, I didn't remember precisely. This however is actually a good opportunity to lay out an example URDF that has a couple of custom parameters and we could show how these are appearing for a SystemInterface implementation. Perhaps a next ticket? I'd be happy to merge this as-is